dali/imgcodec/decoders/nvjpeg2k.h

```diff
@@ -0,0 +1,145 @@
+
+  using BatchParallelDecoderImpl::CanDecode;
+  bool CanDecode(ImageSource *in, DecodeParams opts, const ROI &roi) {
+    return !roi && (opts.dtype == DALI_UINT8);
+  }
+
+  using BatchParallelDecoderImpl::DecodeImplTask;
+  DecodeResult DecodeImplTask(int thread_idx,
+                              cudaStream_t stream,
+                              SampleView<GPUBackend> out,
+                              ImageSource *in,
+                              DecodeParams opts,
+                              const ROI &roi) override;
+
+  void SetParam(const char *name, const any &value) {
+    if (strcmp(name, "nvjpeg2k_device_memory_padding") == 0) {
+      nvjpeg2k_device_memory_padding_ = any_cast<size_t>(value);
+    } else if (strcmp(name, "nvjpeg2k_host_memory_padding") == 0) {
+      nvjpeg2k_host_memory_padding_ = any_cast<size_t>(value);
+    }
+  }
+
+  any GetParam(const char *name) const override {
+    if (strcmp(name, "nvjpeg2k_device_memory_padding") == 0) {
+      return nvjpeg2k_device_memory_padding_;
+    } else if (strcmp(name, "nvjpeg2k_host_memory_padding") == 0) {
+      return nvjpeg2k_host_memory_padding_;
+    } else {
+      DALI_FAIL("Unrecognized param name");
+    }
+  }
+
+ private:
+  struct Context {
+    uint8_t bpp;
+    TensorShape<> shape;
+    int req_nchannels;
+    bool needs_processing;
+    DALIDataType pixel_type;
+    size_t pixel_size;
+    int64_t pixels_count;
+    int64_t comp_size;
+
+    NvJpeg2kDecodeState *nvjpeg2k_decode_state;
+    NvJpeg2kStream *nvjpeg2k_stream;
+    CUDAEvent *decode_event;
+    CUDAStreamLease *cuda_stream;
+  };
+
+  struct ThreadResources {
+    ThreadResources() {}
+    ThreadResources(const NvJpeg2kHandle &nvjpeg2k_handle,
+                    size_t device_memory_padding, int device_id)
+    : nvjpeg2k_decode_state(nvjpeg2k_handle)
+    , intermediate_buffer()
+    , nvjpeg2k_stream(NvJpeg2kStream::Create())
+    , decode_event(CUDAEvent::Create(device_id))
+    , cuda_stream(CUDAStreamPool::instance().Get(device_id)) {
+      intermediate_buffer.resize(device_memory_padding / 8);
+      CUDA_CALL(cudaEventRecord(decode_event, cuda_stream));
+    }
+
+    NvJpeg2kDecodeState nvjpeg2k_decode_state;
+    DeviceBuffer<uint8_t> intermediate_buffer;
+    NvJpeg2kStream nvjpeg2k_stream;
+    CUDAEvent decode_event;
+    CUDAStreamLease cuda_stream;
+  };
+
+  // TODO(staniewzki): remove default values
+  size_t nvjpeg2k_device_memory_padding_ = 256;
+  size_t nvjpeg2k_host_memory_padding_ = 256;
+
+  bool ParseJpeg2000Info(ImageSource *in, DecodeParams opts, Context *ctx);
+  bool DecodeJpeg2000(ImageSource *in, uint8_t *out, DecodeParams opts, Context *ctx);
+  bool ConvertData(void *in, uint8_t *out, DecodeParams opts, Context *ctx);
+
+  NvJpeg2kHandle nvjpeg2k_handle_{};
+  nvjpeg2kDeviceAllocator_t nvjpeg2k_dev_alloc_;
+  nvjpeg2kPinnedAllocator_t nvjpeg2k_pin_alloc_;
+  std::vector<ThreadResources> thread_resources_;
```

Comment:
```suggestion
  std::vector<PerThreadResources> per_thread_resources_;
```
would read better I think. Otherwise it makes me think it's something about threading.